idea-plugin/src/main/kotlin/arrow/meta/ide/MetaRegistrar.kt

```diff
@@ -34,17 +45,43 @@ private val metaPlugin = IdeMetaPlugin()
  * This extension registers a MetaPlugin for a given project.
  */
 private val metaProjectRegistrar: ProjectLifecycleListener
-  get() = object : ProjectLifecycleListener, Disposable {
-    override fun projectComponentsInitialized(project: Project) {
-      val LOG = Logger.getInstance("#arrow.metaProjectRegistrarInProject:${project.name}")
+  get() = projectLifecycleListener(
+    initialize = { project ->
+      val LOG = Logger.getInstance("#arrow.metaProjectRegistrarForProject:${project.name}")
       LOG.info("beforeProjectLoaded:${project.name}")
       val start = System.currentTimeMillis()
       val configuration = CompilerConfiguration()
       metaPlugin.registerMetaComponents(project, configuration)
       LOG.info("beforeProjectLoaded:${project.name} took ${System.currentTimeMillis() - start}ms")
+    },
+    dispose = {
+      // TODO: make sure that all registered extensions are disposed
     }
+  )
 
-    override fun dispose() {
-      // TODO: make sure that all registered extensions are disposed
+/**
+ * [quoteLifecycleRegistrar] addresses ide lifecycle specific manipulates utilizing the [QuoteSystemService], [QuoteCache] and [QuoteHighlightingCache].
+ */
+val quoteLifecycleRegistrar: ProjectLifecycle
+  get() = projectLifecycleListener(
+    // the usual registration should be in `beforeProjectOpened`, but this is only possible when #446 is unlocked
+    initialize = { project: Project ->
+      project.quoteConfigs()?.let { (system, cache) ->
+        initializeQuotes(project, system, cache)
+      }
+    },
+    afterProjectClosed = { project: Project ->
+      project.quoteConfigs()?.let { (quoteSystem, cache) ->
+        try {
+          quoteSystem.context.cacheExec.safeAs<BoundedTaskExecutor>()?.shutdownNow()
+        } catch (e: Exception) {
+          LOG.warn("error shutting down pool", e)
+        } finally {
+          cache.clear()
+          project.getService(QuoteHighlightingCache::class.java)?.map { // resets the highlighter
```

Comment:
Is this needed? The project is closed here

Reply:
Yes to reset the lifecycle of the Highlighter.

Reply:
I removed it